### PR TITLE
Feature/gitlab project configuration

### PR DESCRIPTION
**Pull request by suukit**
##### SUMMARY
Enhanced existing community.general.gitlab_project to handle additional gitlab properties:
- allow_merge_on_skipped_pipeline
- only_allow_merge_if_all_discussions_are_resolved
- only_allow_merge_if_pipeline_succeeds
- packages_enabled
- remove_source_branch_after_merge
- squash_option

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
gitlab_project

##### ADDITIONAL INFORMATION
The current gitlab modules within this collection only partially covert the features and parameters of the gitlab API. To be able to get a fully automated repo-management this has to be enhanced. This is my first enhancement with some simple parameters to let me understand the collaboration workflow and to get in contact.  As we currently enhance the modules to allow for repo mirroring, configuration of integrations I would provide more enhancements within the next weeks if wanted.

### CODE REVIEW

File: changelogs/fragments/3001-enhance_gitlab_module.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - enhanced  gitlab_project o manage attributes allow_merge_on_skipped_pipeline, only_allow_merge_if_all_discussions_are_resolved, only_allow_merge_if_pipeline_succeeds, packages_enabled, remove_source_branch_after_merge, squash_option
```
Comment (felixfontein) [inline changelogs/fragments/3001-enhance_gitlab_module.yml:2]:
```suggestion
  - gitlab_project - add new options ``allow_merge_on_skipped_pipeline``, ``only_allow_merge_if_all_discussions_are_resolved``, ``only_allow_merge_if_pipeline_succeeds``, ``packages_enabled``, ``remove_source_branch_after_merge``, ``squash_option`` (https://github.com/ansible-collections/community.general/pull/3002).
```

File: plugins/modules/source_control/gitlab/gitlab_project.py [hunk 2]
```diff
@@ -114,6 +114,50 @@
       - Used to create a personal project under a user's name.
     type: str
     version_added: "3.3.0"
+  allow_merge_on_skipped_pipeline:
+    description:
+      - allow merge when skipped pipelines exist
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project.py:119]:
```suggestion
      - Allow merge when skipped pipelines exist.
```
Descriptions must always be proper sentences. Please also adjust the other ones.

File: plugins/modules/source_control/gitlab/gitlab_project.py [hunk 3]
```diff
@@ -114,6 +114,50 @@
...
+    type: bool
+    required: false
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project.py:121]:
```suggestion
```
(same below)

File: plugins/modules/source_control/gitlab/gitlab_project.py [hunk 4]
```diff
@@ -308,6 +358,12 @@ def main():
         state=dict(type='str', default="present", choices=["absent", "present"]),
         lfs_enabled=dict(default=False, type='bool'),
         username=dict(type='str'),
+        allow_merge_on_skipped_pipeline=dict(default=False, type='bool'),
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project.py:361]:
```suggestion
        allow_merge_on_skipped_pipeline=dict(type='bool', default=False),
```
Please specify `type` before `default`.

File: plugins/modules/source_control/gitlab/gitlab_project.py [hunk 5]
```diff
@@ -396,7 +458,14 @@ def main():
...
                                                 "visibility": visibility,
                                                 "import_url": import_url,
-                                                "lfs_enabled": lfs_enabled}):
+                                                "lfs_enabled": lfs_enabled,
+                                                "allow_merge_on_skipped_pipeline": allow_merge_on_skipped_pipeline,
+                                                "only_allow_merge_if_all_discussions_are_resolved": only_allow_merge_if_all_discussions_are_resolved,
+                                                "only_allow_merge_if_pipeline_succeeds": only_allow_merge_if_pipeline_succeeds,
+                                                "packages_enabled": packages_enabled,
+                                                "remove_source_branch_after_merge": remove_source_branch_after_merge,
+                                                "squash_option": squash_option
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project.py:467]:
```suggestion
                                                "squash_option": squash_option,
```
Makes the diff of the next change smaller :)

File: plugins/modules/source_control/gitlab/gitlab_project.py [hunk 6]
```diff
@@ -114,6 +114,50 @@
...
+  packages_enabled:
+    description:
+      - enable github package repository
+    type: bool
+    required: false
+    default: false
+    version_added: "3.4.0"
+  remove_source_branch_after_merge:
+    description:
+      - remove the source branch agter merge
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project.py:147]:
```suggestion
      - Remove the source branch after merge.
```

File: plugins/modules/source_control/gitlab/gitlab_project.py [hunk 7]
```diff
@@ -114,6 +114,50 @@
...
+    required: false
+    default: false
+    version_added: "3.4.0"
+  squash_option:
+    description:
+      - squash commits when merging
+    type: str
+    required: false
+    default: 'default_off'
+    choices: ["never","always","default_off", "default_on"]
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project.py:158]:
```suggestion
    choices: ["never", "always", "default_off", "default_on"]
```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution. I've added a first few comments.

A problem with this is that you set explicit defaults for all the new options, which means that if someone uses the module right now (to manage other aspects) and manually already set one of the options your PR adds to another value, the module will suddenly start modifying that option. I think it would be better to *not* have defaults. The change detection already ignores options with value `None` (see `updateProject`). I'm not sure what happens with such options in `createProject` though, that's something you'll have to check.

**Comment by suukit** [comment]
@felixfontein: thank you for your review. I changed the default behavior of the new properties as requested and resolved the other issues.
I tested the following use-cases:
1. Project creation without new parameters: behaves like without this patch
2. Project creation with some/all new parameters: whatever new parameter is used will be set in the project, others get gitlab default values
3. Project update without new parameters: like without patch, even if properties in gitlab where changed using UI now change
4. Project update with new parameters: parameters set to wanted value
5. Project update after using new parameters without new parameters: no change, previously set values remain

File: plugins/modules/source_control/gitlab/gitlab_project.py [hunk 8]
```diff
@@ -254,6 +293,17 @@ def createProject(self, namespace, arguments):
 
         return project
 
+    '''
+    @param arguments Attributes of the project
+    '''
+    def getOptionsWithValue(self, arguments):
+        retArguments = dict()
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project.py:300]:
```suggestion
        ret_arguments = dict()
```
(same for the other usages)

File: plugins/modules/source_control/gitlab/gitlab_project.py [hunk 9]
```diff
@@ -308,6 +358,12 @@ def main():
         state=dict(type='str', default="present", choices=["absent", "present"]),
         lfs_enabled=dict(default=False, type='bool'),
         username=dict(type='str'),
+        allow_merge_on_skipped_pipeline=dict(type='bool'),
+        only_allow_merge_if_all_discussions_are_resolved=dict(type='bool'),
+        only_allow_merge_if_pipeline_succeeds=dict(type='bool'),
+        packages_enabled=dict(type='bool'),
+        remove_source_branch_after_merge=dict(type='bool'),
+        squash_option=dict(type='str', choices=["never", "always", "default_off", "default_on"]),
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project.py:366]:
```suggestion
        squash_option=dict(type='str', choices=['never', 'always', 'default_off', 'default_on']),
```

File: plugins/modules/source_control/gitlab/gitlab_project.py [hunk 10]
```diff
@@ -114,6 +114,38 @@
...
     type: str
     version_added: "3.3.0"
+  allow_merge_on_skipped_pipeline:
+    description:
+      - Allow merge when skipped pipelines exist.
+    type: bool
+    version_added: "3.4.0"
+  only_allow_merge_if_all_discussions_are_resolved:
+    description:
+      - All discussions on a MR have to be resolved.
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project.py:124]:
```suggestion
      - All discussions on a merge request (MR) have to be resolved.
```

File: plugins/modules/source_control/gitlab/gitlab_project.py [hunk 11]
```diff
@@ -114,6 +114,38 @@
...
+    type: bool
+    version_added: "3.4.0"
+  only_allow_merge_if_pipeline_succeeds:
+    description:
+      - Only allow merges if pipeline succeeded.
+    type: bool
+    version_added: "3.4.0"
+  packages_enabled:
+    description:
+      - Enable github package repository.
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project.py:134]: Shouldn't this be GitLab instead of Github?

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@suukit thanks a lot for your contribution!

**Pull request merged**